allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+        # arguments to construct this class in the first place.
+        trainer = trainer_builder.construct(model=model, data_loader=data_loader)
+
+        super().__init__(serialization_dir=serialization_dir, model=model, trainer=trainer)
+
+        self.dataset = dataset
+        self.data_loader_builder = data_loader_builder
+        self.trainer_builder = trainer_builder
+        self.batch_weight_key = batch_weight_key
+        self.cross_validator = cross_validator
+        self.retrain = retrain
+
+    @overrides
+    def run(self) -> Dict[str, Any]:
+        metrics_by_fold = []
+
+        n_splits = self.cross_validator.get_n_splits(self.dataset)
+
+        for fold_index, (train_indices, test_indices) in enumerate(
+            self.cross_validator(self.dataset)
+        ):
+            logger.info(f"Fold {fold_index}/{n_splits - 1}")
+
+            serialization_dir = os.path.join(self.serialization_dir, f"fold_{fold_index}")
+            os.makedirs(serialization_dir, exist_ok=True)
+
+            train_dataset = Subset(self.dataset, train_indices)
+            test_dataset = Subset(self.dataset, test_indices)
+
+            train_data_loader = self.data_loader_builder.construct(dataset=train_dataset)
+            test_data_loader = self.data_loader_builder.construct(dataset=test_dataset)
+
+            model = copy.deepcopy(self.model)
+            subtrainer = self.trainer_builder.construct(
+                serialization_dir=serialization_dir, data_loader=train_data_loader, model=model
+            )
+
+            fold_metrics = subtrainer.train()
+
+            for metric_key, metric_value in training_util.evaluate(
+                model,
+                test_data_loader,
+                subtrainer.cuda_device,
+                batch_weight_key=self.batch_weight_key,
+            ).items():
+                if metric_key in fold_metrics:
+                    fold_metrics[f"test_{metric_key}"] = metric_value
+                else:
+                    fold_metrics[metric_key] = metric_value
+
+            common_util.dump_metrics(
+                os.path.join(subtrainer._serialization_dir, "metrics.json"), fold_metrics, log=True
+            )
+
+            metrics_by_fold.append(fold_metrics)
+
+        metrics = {}
+
+        for metric_key, fold_0_metric_value in metrics_by_fold[0].items():
+            if isinstance(fold_0_metric_value, float):
+                average = Average()
+                for fold_index, fold_metrics in enumerate(metrics_by_fold):
+                    metric_value = fold_metrics[metric_key]
+                    metrics[f"fold{fold_index}_{metric_key}"] = metric_value
+                    average(metric_value)
+                metrics[f"average_{metric_key}"] = average.get_metric()
+            else:
+                for fold_index, fold_metrics in enumerate(metrics_by_fold):
+                    metrics[f"fold{fold_index}_{metric_key}"] = fold_metrics[metric_key]
```

Comment:
You could do this for float values as well, right? Basically, do the `else` block for everything, but for floats, also report an average? Maybe add an stddev/min/max, so that it's easy to see if there us an outlier skewing the mean?